File: src/main/java/framework/ftc/cobaltforge/Injector.java
package framework.ftc.cobaltforge;

import framework.ftc.cobaltforge.exceptions.IncompatibleInjectionException;

import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by Dummyc0m on 9/28/16.
 */
class Injector {
    static CobaltForge runningCobalt;
    private static Injector ourInstance = new Injector();
    private Map<Class<Object>, Object> injectableMap = new HashMap<Class<Object>, Object>();

    private Injector() {
    }

    static Injector getInstance() {
        return ourInstance;
    }

    private Object getInjectable(Class<Object> clazz) throws IllegalAccessException, InstantiationException {
        return injectableMap.containsKey(clazz) ? injectableMap.get(clazz) : clazz.newInstance();
    }

    private void injectObject(Object object) {
        Field[] fields = object.getClass().getDeclaredFields();
        for (Field field : fields) {
            field.setAccessible(true);
            Annotation[] annotations = field.getAnnotations();
            for (Annotation annotation : annotations) {
                Class clazz = annotation.annotationType();
                if (clazz == Inject.class) {
                    injectField(field, object);
                }
            }
        }
    }

    @SuppressWarnings("unchecked")
    void injectField(Field field, Object directive) {
        try {
            Class clazz = field.getType();
            if (clazz == CobaltForge.class) {
                System.out.println("Injected CobaltForge");
                field.set(directive, runningCobalt);
            } else {
                Object object = getInjectable(clazz);
                injectObject(object);
                System.out.println("Injected " + object);
                field.set(directive, object);
            }
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            throw new IncompatibleInjectionException(e);
        } catch (ClassCastException e) {
            throw new IncompatibleInjectionException(field, Object.class);
        }
    }
}

File: src/main/java/framework/ftc/cobaltforge/Injector.kt
package framework.ftc.cobaltforge

import framework.ftc.cobaltforge.exceptions.IncompatibleInjectionException
import java.lang.reflect.Field
import java.util.*

/**
 * Created by Dummyc0m on 9/28/16.
 */
internal object Injector {
    private val injectableMap = HashMap<Class<*>, Any>()
    var runningCobalt: CobaltForge? = null

    @Throws(IllegalAccessException::class, InstantiationException::class)
    private fun getInjectable(clazz: Class<*>): Any? {
        return if (injectableMap.containsKey(clazz)) injectableMap[clazz] else clazz.newInstance()
    }

    private fun injectObject(obj: Any) {
        val fields = obj.javaClass.declaredFields
        for (field in fields) {
            field.isAccessible = true
            val annotations = field.annotations
            for (annotation in annotations) {
                val clazz = annotation.annotationClass
                if (clazz == Inject::class.java) {
                    injectField(field, obj)
                }
            }
        }
    }

    @SuppressWarnings("unchecked")
    fun injectField(field: Field, directive: Any) {
        try {
            val clazz = field.type
            if (clazz == CobaltForge::class.java) {
                println("Injected CobaltForge")
                field.set(directive, runningCobalt)
            } else {
                val obj = getInjectable(clazz)
                injectObject(obj!!)
                println("Injected " + obj)
                field.set(directive, obj)
            }
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InstantiationException) {
            throw IncompatibleInjectionException(e)
        } catch (e: ClassCastException) {
            throw IncompatibleInjectionException(field, Any::class.java)
        }

    }
}
